Implements #18 issue/feature, the "to _radix_" command:

mgaggero opened the pull request:
Implements the "to radix" command to quickly and temporary convert a number or the result of an expression to a different radix (for example when working with memory address and segments sizes):

```
cork> 0xCAFE to dec
51966
cork> 0xCAFE + 2 * 8 * 0x20 to dec
52478
cork> 0xCAFE + 2 * 8 * 0x20
0xccfe
cork> ans to bin
0b1100110011111110
cork> 1234 to hex
0x4d2
cork> (0xCAFE + 2 * 8 * 0x20) to oct
0o146376
```
![cork_to_radix_example](https://user-images.githubusercontent.com/6576853/196555798-20dce055-729e-4c53-923e-cbb4ab2cf25f.png)

mgaggero committed:
Implements #18 issue/feature, the "to _radix_" command:
  cork> 0xCAFE to dec
  51966

RedDocMD commented on src/expression.peg:
```diff
@@ -21,8 +21,9 @@ expr = { term ~ (operation ~ term)* }
 term = _{ ans | number | "(" ~ expr ~ ")" }
 
 set_directive = @{ "set " ~ (word ~ " ")* ~ word }
+tor_directive = { term ~ (operation ~ term)* ~ "to " ~ word }
```
I think this can also be written as:
```
tor_directive = { expr ~ "to " ~ radix }
radix = { "dec" | "oct" | "hex" | "bin" }
```
IMO, this is cleaner and expresses the intent better.

mgaggero replied:
Ok, it's more Pest's idiomatic (I've never used Pest before).

RedDocMD commented on src/expression.rs:
```diff
@@ -98,13 +99,44 @@ impl Index<usize> for SetDirective {
+#[derive(Debug, PartialEq, Eq)]
+pub struct ConvDirective {
+    expr: Expr,
+    radix: String
```
This type is better suited to be `FormatRadix`.

mgaggero replied:
Ok.

RedDocMD commented on src/expression.rs:
```diff
@@ -98,13 +99,44 @@ impl Index<usize> for SetDirective {
+        eval::eval_expr(&self.expr, ans)
+    }
+
+    pub fn radix(&self) -> Result<FormatRadix, CorkError> {
```
This function will become trivial if the type of `radix` is `FormatRadix`. Also the `Result` will be un-neccessary.

mgaggero replied:
Ok.

RedDocMD commented on src/expression.rs:
```diff
@@ -124,6 +156,14 @@ fn parse_comm(pair: Pair<Rule>) -> Command {
         }),
+        Rule::tor_directive => {
+            let to_radix = pair.as_str().split(' ').map(String::from)
+                .collect::<Vec<String>>().pop().unwrap();
```
You don't need to `collect` into a `Vec` and then `pop`. You can simply call `take` after the `map` - since `map` also returns an Iterator. The final `unwrap` will of-course be  necessary.

mgaggero replied:
Ok, now the line has been totally changed. I've finally found enough time to deepen the knowledge on Pest and to modify the code. I've applied `rustfmt` on `src/expression.rs` only since in the other files it was changing lines not edited by me (maybe more readable for the author in the previous format). Thanks for the suggestions.

RedDocMD reviewed (changes requested):
Nice work!
I have made a few comments. Please look into them.
Also, I think you may need to run your code through `rustfmt`.

mgaggero committed:
- refactoring of the grammar, following https://github.com/RedDocMD/cork/pull/19#discussion_r1002013117

mgaggero committed:
Specific conversion error is no more required with grammar refactoring
(pest will issue an appropriate error instead).

mgaggero committed:
Increases code readability. Thanks to @RedDocMD:
- https://github.com/RedDocMD/cork/pull/19#discussion_r1002026708
- https://github.com/RedDocMD/cork/pull/19#discussion_r1002027389
- https://github.com/RedDocMD/cork/pull/19#discussion_r1002028742

mgaggero committed:
Run rustfmt on src/expression.rs.

mgaggero reviewed (commented):
Commit 8b94b9e implements the requested changes.

RedDocMD commented:
Thank you for the changes!
I think a test or two would not be amiss. In particular, you can see a given string is actually parsed into a `Command::Convert`, and if invalid bases are rejected.

mgaggero committed:
Adds tests for inline result's radix conversion ('to radix' command).

mgaggero commented:
> Thank you for the changes! I think a test or two would not be amiss. In particular, you can see a given string is actually parsed into a `Command::Convert`, and if invalid bases are rejected.

You're right, sure, I forgot the tests. Added in the last commit.

mgaggero committed:
Updates README.md: adds temporary format conversion.

RedDocMD commented:
Thanks a lot! I think I will merge it now, since you have covered all bases pretty nicely.
